docs/api_docs.rst

```diff
@@ -0,0 +1,316 @@
+of the image along with paired mask matrices indicating where perturbations have occurred and to what amount.
+
+Implementations should impart no side effects upon the input image.
+
+Immediate candidates for implementation of this interface are occlusion-based saliency algorithms [3] that perform
+perturbations on image pixels.
+
+Interface: PerturbImage
+------------------------
+.. autoclass:: xaitk_saliency.interfaces.perturb_image.PerturbImage
+   :members:
+
+-----------------------------------------
+Image Occlusion via Perturbation Masks
+-----------------------------------------
+
+A common intermediate step in this process is applying the generated perturbation masks to imagery to produce occluded
+images.
+We provide two utility functions as baseline implementation to perform this step:
+
+* ``xaitk_saliency.utils.masking.occlude_image_batch`` - performs the transformation as a batch operation
+
+* ``xaitk_saliency.utils.masking.occlude_image_streaming`` - performs the transformation in a streaming method with
+optional parallelization in that streaming
+
+While the batch version is simpler and in many cases the faster of the two versions, the streaming version may be more
+applicable to large image masks or when a great deal of masks are being input, where in such cases the batch version
+would exceed available memory.
+
+.. autofunction:: xaitk_saliency.utils.masking.occlude_image_batch
+
+.. autofunction:: xaitk_saliency.utils.masking.occlude_image_streaming
+
+
+----------------------------
+Saliency Heatmap Generation
+----------------------------
+
+These interfaces comprise a family of siblings that all perform a similar transformation, but requiring different
+standard inputs.
+There is no standard to rule them all without being so abstract that it would break the concept of interface
+abstraction, or the ability to substitute any arbitrary implementations of the interface without interrupting successful
+execution.
+Each interface is intended to handle different black-box outputs from different algorithmic categories.
+In the future, as additional algorithmic categories are identified for which saliency map generation is applicable,
+additional interfaces may be defined and added to this initial repertoire.
+
+Interface: GenerateDescriptorSimilaritySaliency
+-----------------------------------------------
+
+This interface proposes that implementations require externally generated feature-vectors for two reference images
+between which we are trying to discern the feature-space saliency.
+This also requires the feature-vectors for perturbed images as well as the masks of the perturbations as would be
+output from a ``PerturbImage`` implementation.
+We expect perturbations to be relative to the second reference image feature-vector.
+
+An immediate candidate implementation for this interface is the SBSM algorithm [1].
+
+.. autoclass:: xaitk_saliency.interfaces.gen_descriptor_sim_sal.GenerateDescriptorSimilaritySaliency
```

Comment:
Looks like the `nFeats` shape in the comment for ref_descr_2 has extra quotation marks? See [here](https://xaitk-saliency--51.org.readthedocs.build/en/51/api_docs.html#xaitk_saliency.interfaces.gen_descriptor_sim_sal.GenerateDescriptorSimilaritySaliency)

This probably needs to be changed in the actual source code, so we are using autodoc here.

Reply:
should I fix this is as part of this PR?

Reply:
I think you might as well fix this since it is so minor.

Reply:
done, also fixed a couple other typos